nano/secure/ledger.cpp

```diff
@@ -538,11 +577,17 @@ void ledger_processor::open_block (nano::open_block const & block_a)
 	result.code = existing ? nano::process_result::old : nano::process_result::progress; // Have we seen this block already? (Harmless)
 	if (result.code == nano::process_result::progress)
 	{
-		auto source_missing (!ledger.store.block_exists (transaction, nano::block_type::send, block_a.hashables.source) && !ledger.store.block_exists (transaction, nano::block_type::state, block_a.hashables.source));
-		result.code = source_missing ? nano::process_result::gap_source : nano::process_result::progress; // Have we seen the source block? (Harmless)
+		// Validate block if not verified outside of ledger
+		if (result.verified != nano::signature_verification::valid)
+		{
+			result.code = validate_message (block_a.hashables.account, hash, block_a.signature) ? nano::process_result::bad_signature : nano::process_result::progress; // Is the signature valid (Malformed)
+		}
 		if (result.code == nano::process_result::progress)
 		{
-			result.code = nano::validate_message (block_a.hashables.account, hash, block_a.signature) ? nano::process_result::bad_signature : nano::process_result::progress; // Is the signature valid (Malformed)
+			assert (!validate_message (block_a.hashables.account, hash, block_a.signature));
+			result.verified = nano::signature_verification::valid;
+			auto source_missing (!ledger.store.block_exists (transaction, nano::block_type::send, block_a.hashables.source) && !ledger.store.block_exists (transaction, nano::block_type::state, block_a.hashables.source));
```

Comment:
I think it may be more readable to factor out the boolean operands into 2 variables then do something like:
`auto source_missing (not_send_block && not_state_block);`

It could also be done here but because it is a boolean OR || it means the right operand is called regardless, so maybe just refactor the whole boolean expression `block_a.hashables.source) || ledger.store.block_exists (transaction, nano::block_type::state, block_a.hashables.source` into a variable: https://github.com/nanocurrency/nano-node/blob/ab62b094e5d81a598da3f8b7842d10d2f4efddd7/nano/secure/ledger.cpp#L528

Reply:
Thanks! Made new nano::block_store::source_exists () function for this